feat: 글 및 댓글 작성자 이름 앞에 프로필 사진 보여주기 구현 [ISSUE-454]

yelimkim98 opened the pull request:
**이슈 번호**

resolved: #454 

**작업 내용**
![image](https://user-images.githubusercontent.com/28949340/93711148-f7373000-fb86-11ea-91fc-8abc8b806144.png)

1. 글 및 댓글 작성자 이름 앞에 프로필 사진 보여주기 구현
2. 자잘한 버그 해결 (/home 가면 alert 뜨고 나서 로그인 페이지 가지는 현상 등등)

**테스트 작성 여부**

- [ ] Test case

**주의 사항**

yelimkim98 committed:
feat: 글 상세 화면에서 작성자 프로필 사진 작은 동그라미로 보여주기 구현

yelimkim98 committed:
feat: 댓글 작성자 프로필 조그만 동그라미로 보여주기 및 프로필 페이지로 link

 - 댓글 왼쪽에 댓글 작성자의 프로필 사진을 보여줌.
 - 프로필 사진을 클릭하면 댓글 작성자의 프로필을 볼 수 있음.

yelimkim98 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into develop

 Conflicts:
	front-end/legeno-around-here/src/components/pages/post/PostDetail.js

yelimkim98 committed:
refactor: 글 작성자 이름과 프사 사이 간격을 조금 넓힘

yelimkim98 committed:
fix: 프론트 로그인 안한 상태에서 /home 접근시 예기치못한 에러 문제 해결

 - 비동기 관련하여 문제가 발생하는 듯 했음.
   findAllSimpleSectors 메서드에서 response가 null 인데 response.data 에
   접근하려 하는 경우가 있었고,
   HomeTopBar.js 에서 findAllSimpleSectors 메서드를 호출하는 부분에
   undefined.unshift(..) 이렇게
   undefined 객체에서 메서드를 호출하는 현상이 있었음.
 - alert 가 뜬 다음 로그인페이지로 리다이렉트 되는 문제 해결함

aegis1920 commented on front-end/legeno-around-here/src/components/pages/home/HomeTopBar.js:
```diff
@@ -31,9 +31,16 @@ const HomeTopBar = ({ setter, sectorId, history }) => {
-      const foundSimpleSectors = await findAllSimpleSectors(accessToken, history);
-      await foundSimpleSectors.unshift(SIMPLE_ALL_SECTOR);
-      await setSimpleSectors(foundSimpleSectors);
+      // const foundSimpleSectors = await findAllSimpleSectors(accessToken, history);
+      // await foundSimpleSectors.unshift(SIMPLE_ALL_SECTOR);
+      // await setSimpleSectors(foundSimpleSectors);
```
주석 처리한 부분 지워도 될 것 같아요!!

yelimkim98 replied:
앗 ! 수정했습니다!!

aegis1920 commented on front-end/legeno-around-here/src/components/pages/post/CommentItem.js:
```diff
@@ -1,13 +1,59 @@
+    }
+    return <LinkWithoutStyle
+      className={classes.creatorProfilePhoto}
+      to={isMyComment? '/users/me' : '/users/' + comment.creator.id}
```
? 앞에 띄어쓰기가 안 되어있는 것 같네요??

yelimkim98 replied:
수정했습니다!!

aegis1920 commented on front-end/legeno-around-here/src/components/pages/post/PostDetail.js:
```diff
@@ -23,6 +55,9 @@ const PostDetail = ({ post, myInfo, history }) => {
   const [zzang, setZzang] = useState(post.zzang.activated);
   const [zzangCount, setZzangCount] = useState(post.zzang.count);
   const isMyPost = post && myInfo && post.creator.id === myInfo.id;
+  const authorProfilePhotoUrl= post.creator.image ? post.creator.image.url : DEFAULT_IMAGE_URL;
```
여기도 = 앞에 띄어쓰기~

yelimkim98 replied:
수정했습니다!!

aegis1920 commented on front-end/legeno-around-here/src/components/pages/post/PostDetail.js:
```diff
@@ -62,28 +97,40 @@ const PostDetail = ({ post, myInfo, history }) => {
   };
 
-  const findCreatorLink = () => {
+  const makeCreatorNameUi = () => {
```
메서드명에 Ui를 붙인 이유가 있을까요?

yelimkim98 replied:
이 메서드가 JSX 을 리턴하고 JSX 안에서 쓰이는데요!
`findCreatorLink ` 이 이름을 쓰다보니까
자꾸 URL 문자열을 리턴하는줄로 햇갈려서 불편하더라구요!
그래서 이렇게 하긴 했는데 더 좋은 네이밍을 고민해볼 필요가 있을 것 같네요..!!

yelimkim98 replied:
`createCreatorNameJsx`
`makeCreatorNameJsx`
`createCreatorNameTag`
`makeCreatorNameTag`
`creatorNameTag`

몇가지 생각해봤는데... 어떠신가요?! 뭐가 가장 괜찮을까요?! 다른 아이디어를 내주시는것도 좋아요!!

aegis1920 replied:
제 생각에는 그냥 Ui를 빼고 makeCreatorName만 해도 좋을 것 같은데요?

yelimkim98 replied:
그런가요? 알겠습니다!

yelimkim98 replied:
수정했습니다!!

aegis1920 commented on front-end/legeno-around-here/src/components/pages/post/PostDetail.js:
```diff
@@ -147,7 +194,13 @@ const PostDetail = ({ post, myInfo, history }) => {
+        <Comments
+          comments={comments}
+          loading={loading}
+          myUserId={myInfo? myInfo.id : null}
```
할당하는 곳에는 myInfo인데 변수명은 myUser네요?
+ 여기도 앞에 띄어쓰기 부탁합니당ㅎㅎ
+ null을 전달해주는 이유가 있을까요?

yelimkim98 replied:
1
myUserId 에서 myId로 바꿨습니다!
ID라고만 하면 이메일과 햇갈릴까봐 UserId라고 한거였는데 오히려 햇갈릴 것 같네요!

2
띄어쓰기 수정했습니다!

3
null을 전달하는 이유는
myinfo 가 null 일 때 myInfo.id 에 접근하려고 해서
에러가 나는 경우가 있더라구요!
비동기 때문에 일시적으로 myInfo가 null 인 순간이 있는 것 같아요..! (아마)

저렇게 해두면 발생할 수 있는 문제점이
내 댓글에서  댓글작성자 프로필 보기를 했는데
타인 프로필 페이지로 연결될 수 있습니다.
( myId 가 commentCreatorId 와 같은지 확인해서
링크를 users/me 또는 users/commentCreatoeId 로 분기하는 데에 사용되거든요!
내 id가 null이면 댓글작성자 아이디와 항상 다르니까요..!!)
 
그런데 제가 사용해 봤을 때 그런적은 없었고,
실제로 그럴 경우가 있다 하더라도 사용자 입장에서 크리티컬한 문제가 아닐 것 같아서
저렇게 하는게 괜찮아보였습니다!

aegis1920 replied:
저 방식으로 해도 Comments 쪽에 null을 전달해서 Comments 쪽에서 에러가 나지 않나요? 
Comments 컴포넌트 앞에 `myInfo && <Comments ~/>` 를 사용하면 myInfo가 제대로 불러와졌을 때만 보이게 할 수 있어서 더 나은 방식이 될 수도 있을 것 같아요!

yelimkim98 replied:
오호 그런 방법이 있었군요! 수정하겠습니다! 감사합니다!!

aegis1920 reviewed (commented):
수고하셨어요 히히~
몇 가지 궁금한 점 남겼습니다! 확인해주세요 :)

yelimkim98 committed:
refactor: 불필요한 주석 제거

yelimkim98 committed:
refactor: 변수명 및 띄어쓰기 컨벤션 수정

yelimkim98 committed:
refactor: 메서드 네이밍 수정

Hamliet commented on front-end/legeno-around-here/src/components/api/API.js:
```diff
@@ -276,8 +276,6 @@ export const findMyInfo = (accessToken, history) => {
       return userResponse;
     })
     .catch(async (error) => {
-      const errorResponse = error.response.data;
```
redirectLoginWhenUnauthorized외의 에러를 체크하려면 필요해보이는데 지운 이유가 있을까요?

yelimkim98 replied:
이 부분을 전반적으로 없앤 이유는,
저게 있으면 만약 홈화면에 갔다가 로그인 페이지로 리다이렉트 시킬 때
개발자를 위한 alert가 뜨고나서 로그인페이지가 나옵니다..
이게 실 사용자 입장에서는 불편한 상황일 것이기에
이 문제를 해결하고자 했습니다.

로그인 리다이렉트 함수에서 걸러져야하는데 자꾸
Alert가 실행되길래 (이것도 비동기 관련 문제인건지.. 쩝)
그냥 빼버렸습니다.

제가 문제상황을 확인했던 부분에서만 alert를 지웠어야했는데
정보를 얻어오는 api들의 alert를 전반적으로 지웠네요..

글작성, 회원가입처검 CUD 에서만 ALERT 가 필요하다고 생각했습니다.
그리고 alert로 디버깅하는건 (그리고 그게 실서버에까지 올라가는건) 좋지 않다고 생각했습니다..!

Hamliet replied:
특정 api의 경우
    await redirectLoginWhenUnauthorized(error, history);
    const errorResponse = error.response.data;
    alert(errorResponse.errorMessage);
와 같이 순서만 바꾸셨는데 순서를 바꾼것과 완전히 지운것은 어떤 기준으로 하신건가요?

Hamliet replied:
완전히 지우고 redirectLoginWhenUnauthorized 로만 체크하면 403과 500에러 외의 에러에 대해서 처리가 안될 것 같은데요, 모든 에러에 대해서 login으로 redirect 해야한다면 redirectLoginWhenUnauthorized 함수를 수정해야할 것 같습니다.

yelimkim98 replied:
기존에 "undefied" 라고 쓰여진 alert가 자꾸 나타났습니다.
그게 문제라고 생각해서 alert(error.response.data)를 지운거였습니다..

특별한 사용자 input없이 호출되는 api들
(예를들면 홈화면에 들어갈 경우 최신글, 글작성자 닉네임, 부문목록 등,
그 페이지에서 기본적으로 내려받는 정보에 대한 api들)이
alert를 지우는 1차 타겟이었습니다.

이 외에도 사용자 입장에서 의미없어보이는 alert 들을 지우려고 했습니다.
(예를들어 로그인에 실패했을때의 alert는 의미있지만,
만약 어떤 페이지 접속시 에러가 난 경우
alert로 undefined 같은 메세지가 뜬 뒤에 하얀 화면이 나오든
alert없이 하얀화면이 나오든 사용자 입장에선 둘다 똑같이 이상하다고 생각할 것입니다)

그런데 이런 alert가 정말 많았고,
사실 이런 상황은 alert가 있으나 없으나 똑같이 이상한 상황이기때문에 

도중에 그냥 순서만 바꾸는걸로 방향을 바꿨습니다..
(일단 로그인 검사에서 걸리면 alert가 실행되지 않기를 유도한 것인데
제가 비동기에 대한 이해가 부족해서
제 의도와 달리 login redirect 할 때도  alert가 뜰지도 모릅니다..)

저는 alert 부분이 있는것이
403과 500 이외의 에러를 처리해준다고 생각하지 않습니다...
제가 경험했던 모든 비정상적인 상황에서의 alert 내용이 undefined 였고
그러다보니 어느 api 메서드에서 alert 코드가 실행된건지도 알 수 없었습니다.
이러한데 alert가 의미가 있는걸까요??

Hamliet replied:
alert이 있는것이 따로 에러 처리를 하고있진 않지만 에러 여부를 알 수 있어서 필요할 것 같아요! 
만약 현재 상태에서 서버 내부에서 401에러를 내려주면 로그인 페이지로도 이동 안하고 alert도 안뜨기떄문에 사용자와 개발자 모두 에러상황을 인지하지 못할 것 같아요. 
한 예로, 필터를 설정했는데 401에러때문에 글 리스트를 못가져온 상황에서 로그인페이지로 이동하지 않기 때문에 빈 리스트 보여준채로 멈춰있는 상태가 될 텐데요, 이 상황에서 글이 없는건지 에러인건지 알 수가 없지 않을까요?

error.response.data가 undefined라면 error를 콘솔로 찍어서 갖고있는 메시지 경로를 다시 설정해야 할 것 같아요
error를 디버그로 찍어보면 에러 메시지를 갖고있긴 할거에요 (error.message라던지, 이 부분은 확인 필요)

아니면 위에서 말씀드린것처럼 모든 에러에 대해 로그인페이지로 리다이렉트 하는 방법도 있을것 같은데 이 부분은 에러를 어떻게 처리할지에 대해서 팀원들과 더 이야기해보고 정하면 좋을 것 같네요!

yelimkim98 replied:
일단 다시 기존의 alert 코드를 되살렸습니다!

대신 비로그인 유저의 /home 등 접근시 undefined alert가 뜨는 문제를 다른 방식으로 해결했습니다.

catch 블럭에서 로그인 리다이렉트 동작이 수행되면
그 catch 블럭을 끝내도록 하였습니다!
(기존에는
403 확인부터 하도록 순서를 바꿨어도
그 다음 alert가 그냥 실행되더라구요!)

Hamliet commented on front-end/legeno-around-here/src/components/api/API.js:
```diff
@@ -302,8 +300,6 @@ export const findCurrentPostsFromPage = async (page, accessToken, mainAreaId, se
     )
     .then((response) => response.data.content)
     .catch(async (error) => {
-      const errorResponse = error.response.data;
-      alert(errorResponse.errorMessage);
```
여기두요!

Hamliet commented on front-end/legeno-around-here/src/components/api/API.js:
```diff
@@ -318,8 +314,6 @@ export const findMyAwards = async (accessToken, history) => {
     .get(DEFAULT_URL + `/awards/me`, config)
     .then((response) => response.data)
     .catch(async (error) => {
-      const errorResponse = error.response.data;
-      alert(errorResponse.errorMessage);
```
여기두요! :) 
(아래 다른 것들도요)

Hamliet commented on front-end/legeno-around-here/src/components/pages/post/CommentItem.js:
```diff
@@ -1,13 +1,59 @@
+      <ListItem alignItems='flex-start' className={classes.photoAndTextsLayout}>
+        {makeCreatorPhotoUi()}
+        <div className={classes.textsLayout}>
+          <Typography variant='subtitle1'>작성자 : {comment.creator.nickname}</Typography>
```
'작성자 : ' 는 없는것이 깔끔해보입니다!

yelimkim98 replied:
지금 작성자와 본문의 폰트 및 글자색이 똑같아서 일단 남겨두었습니다..!
작성자 이름 색을 회색으로 하거나 글자 크기를 좀 줄이거나 하는 방안으로 바꿀 필요가 있어보입니다!
사실 귀찮아서 안했어요... ㅠㅠ 여기다 얹어서 해볼게요!

Hamliet replied:
글씨체 특성상 전부 Bold처럼 보여서 앱 전반적으로 수정이 필요해보이긴 하네요 😅 
(글 내용, 댓글 내용등은 좀 더 얇은 글씨체로?)

yelimkim98 replied:
'작성자 : ' 을 없애고 작성자 이름을 회색처리했습니다!
글씨체가 Bold 처럼 보이는 부분은 다른 이슈 파는거로 할까요?! ٩(๑꒦ິȏ꒦ິ๑)۶

Hamliet replied:
넵 회의 후에 다른이슈로 해결해도 좋을 것 같아요! :)

Hamliet reviewed (commented):
수고하셨어요 히히! 
코멘트 남겼으니 확인 부탁드려요 😄

yelimkim98 committed:
refactor: 내정보 가져오기 전에 홈화면 댓글목록 안가져오도록 함

yelimkim98 committed:
refactor: 403 응답코드 검사가 빠진 API 함수에 403 응답코드 검사 추가

yelimkim98 committed:
refactor: 비인가된 사용자가 /home 접근시 undefined alert 가 뜨는 문제 해결

로그인 하지 않은 사용자가 /home url에 접근할 경우
"undefined"라는 메세지의 alert 가 4번 연속해서 뜬 뒤
로그인 페이지로 리다이렉트되는 문제가 있었다.

로그인 페이지로 리다이렉트 될 경우 catch 블록을 끝내도록 하여 해결하였다.

yelimkim98 committed:
refactor: 댓글에 '작성자 : '을 지우고 작성자 이름을 회색으로 바꿈

Hamliet reviewed (approved):
고생했어요 히히! Approve 하겠습니다! 😄

aegis1920 reviewed (approved):
피드백 반영하느라 고생했어요 히히! 👍
